platform/lang-impl/src/com/intellij/build/BuildTreeConsoleView.java

```diff
@@ -495,30 +523,25 @@ private Runnable addChildFailureNode(@NotNull ExecutionNode parentNode, @NotNull
     else {
       failures = Collections.singletonList(failure);
     }
-    failureNode.setResult(new FailureResultImpl(failures));
+    ExecutionNode updatedRoot = failureNode.setResult(new FailureResultImpl(failures));
+    if (updatedRoot == null) {
+      updatedRoot = parentNode;
+    }
+    strcutreChanged.add(updatedRoot);
     myConsoleViewHandler.addOutput(failureNode, failure);
     return showErrorIfFirst(failureNode, failure.getNavigatable());
   }
 
-  private void finishChildren(@NotNull ExecutionNode node, @NotNull EventResult result) {
-    if (node.hasEmptyChildren()) return;
-    invokeLater(() -> {
-      for (ExecutionNode child : node.getChildList()) {
-        if (child instanceof ExecutionNode) {
-          ExecutionNode executionChild = (ExecutionNode)child;
-          if (!executionChild.isRunning()) {
-            continue;
-          }
-          invokeLater(() -> {
-            // Need to check again since node could have finished on a later event.
-            if (executionChild.isRunning()) {
-              finishChildren(executionChild, result);
-              executionChild.setResult(result);
-            }
-          });
-        }
+  private static void finishChildren(@NotNull SmartHashSet<ExecutionNode> structureChanged,
+                                     @NotNull ExecutionNode node,
+                                     @NotNull EventResult result) {
+    for (ExecutionNode child : node.getChildList()) {
```

Comment:
This needs to be copied before being iterated over since `child.setResult(result)` may update the list. We learned this from canary builds. I'll update the PR a little later.

Reply:
ok )

Reply:
updated

Reply:
done